nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,286 @@
+	for _, node := range nodes {
+		elem := make(map[string]interface{})
+		elem["id"] = node.ID
+		elem["ip_address"] = node.IPAddress
+		nodeList = append(nodeList, elem)
+	}
+	return d.Set("nodes", nodeList)
+}
+
+func resourceNsxtManagerClusterCreate(d *schema.ResourceData, m interface{}) error {
+	// Call Joincluster function on nodes that are not in the cluster
+	nodes := getClusterNodesFromSchema(d)
+	if len(nodes) == 0 {
+		return fmt.Errorf("Please provide at least one node ip")
+	} else if len(nodes) == 1 {
+		log.Printf("[INFO] Only one node ip provided, skip joining nodes process")
+		return resourceNsxtManagerClusterRead(d, m)
+	}
+	hostIP := getHostIPFromNsxtClient(m)
+	clusterID, certSha256Thumbprint, err := getClusterInfoFromHostNode(d, m)
+	if err != nil {
+		return err
+	}
+
+	for _, guestNode := range nodes {
+		err := joinNodeToCluster(clusterID, certSha256Thumbprint, guestNode, hostIP, d, m)
+		if err != nil {
+			return err
+		}
+	}
+	d.SetId(clusterID)
+	d.Set("cluster_id", clusterID)
+	return resourceNsxtManagerClusterRead(d, m)
+}
+
+func getHostIPFromNsxtClient(client interface{}) string {
+	c := client.(nsxtClients)
+	hostIP := c.Host
+	hostIP = strings.TrimPrefix(hostIP, "https://")
+	return hostIP
+}
+
+func getClusterInfoFromHostNode(d *schema.ResourceData, m interface{}) (string, string, error) {
+	connector := getPolicyConnector(m)
+	client := nsx.NewClusterClient(connector)
+	clusterConfig, err := client.Get()
+	if err != nil {
+		return "", "", fmt.Errorf("Failed to get cluster info from host node: %s", err)
+	}
+	clusterID := *clusterConfig.ClusterId
+	nodes := clusterConfig.Nodes
+	node := nodes[0]
+	apiListenAddr := node.ApiListenAddr
+	certSha256Thumbprint := *apiListenAddr.CertificateSha256Thumbprint
+	return clusterID, certSha256Thumbprint, nil
+}
+
+func getNewNsxtClient(node NsxClusterNode, d *schema.ResourceData, clients interface{}) (interface{}, error) {
+	// The API of joining a new node to cluster needs to be called with the new node's ip,
+	// to do that we initialize a new nsxtClient with new node's ip as its host
+	username := *node.UserName
+	password := *node.Password
+	host := *node.IPAddress
+	if !strings.HasPrefix(host, "https://") {
+		host = fmt.Sprintf("https://%s", host)
+	}
+	c := clients.(nsxtClients)
+	commonConfig := c.CommonConfig
+	newClients := nsxtClients{
+		CommonConfig: commonConfig,
+	}
+	err := configureNewClient(&newClients, &c, host, username, password)
+	if err != nil {
+		return nil, err
+	}
+	return newClients, nil
+}
+
+func configureNewClient(newClient *nsxtClients, oldClient *nsxtClients, host string, username string, password string) error {
+	newClient.Host = host
+	securityCtx := core.NewSecurityContextImpl()
+	oldSecurityCtx := oldClient.PolicySecurityContext
+	oldSecurityCtxData := oldSecurityCtx.GetAllProperties()
+	authScheme, securityContextNeeded := oldSecurityCtxData[security.AUTHENTICATION_SCHEME_ID]
+	if securityContextNeeded {
```

Comment:
Some code here is very similar to configurePolicyConnectorData(), worthwhile to consolidate these if possible

Reply:
I feel like the part that can be put into one function is the part of configuring security context, I have added a function for this part